### PR TITLE
Remove guest mode–require login and open in interactive mode instead

### PR DESCRIPTION
**Pull request by jurgenwerk**
Changes in this PR (behavioural):

1. Removing guest mode - the only way to interact with a card is by being logged in and entering interactive mode. We are removing guest mode because we will find a better way to share cards in the future
2. Removing the concept of toggling operator mode - because of (1), we are now always in operator mode
3. Visiting a card directly using its URL requires the viewer to be logged in and then it opens the attempted card by putting it on the stack in interactive mode

Changes in this PR (technical):
1. Operator mode state query params were moved from card route to the index route (or controller)
2. Card route, which was previously responsible for loading the card and preparing everything necessary to show the card in guest mode and providing mechanisms to open it in operator mode, was effectively reduced to simply getting the attempted card URL and redirecting to the index route such that it puts the attempted card in the operator mode state query param–operator mode will then guarantee 2 things; that the user is logged in, and the stack opens in interactive mode as dictated by the operator mode state query param

To be able to fully test this (i.e. the case where user copies the card url and tries to visit it) this PR first has to be merged:   https://github.com/cardstack/boxel/pull/1615. Even though these PRs are independent from each other they will probably result in merge conflicts so we'll see what gets merged first.

### CODE REVIEW

**Comment by backspace** [comment]
I’ll try this locally but in the preview deployment this is happening for me:

screencast 2024-10-01 07-53-54

**Comment by habdelra** [comment]
> I’ll try this locally but in the preview deployment this is happening for me:
> 
> screencast 2024-10-01 07-53-54 [ screencast 2024-10-01 07-53-54 ](https://private-user-images.githubusercontent.com/43280/372475560-064ab288-15ab-467e-927f-309efba602e5.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc3ODgxOTAsIm5iZiI6MTcyNzc4Nzg5MCwicGF0aCI6Ii80MzI4MC8zNzI0NzU1NjAtMDY0YWIyODgtMTVhYi00NjdlLTkyN2YtMzA5ZWZiYTYwMmU1LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDAxVDEzMDQ1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZmNzJkMDkyZDJmYjFiMjRhZmM4ODdlYTUwMDY1MzMyZTc3YWVlMTY0ODAyOTE1NjBjYmQ2OWU0N2ZjMzZlZDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Flckf3Jd-yXbU2jvjdiqwnPC7QNIlHJ4fD7N2bF5Z5I) [ ](https://private-user-images.githubusercontent.com/43280/372475560-064ab288-15ab-467e-927f-309efba602e5.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc3ODgxOTAsIm5iZiI6MTcyNzc4Nzg5MCwicGF0aCI6Ii80MzI4MC8zNzI0NzU1NjAtMDY0YWIyODgtMTVhYi00NjdlLTkyN2YtMzA5ZWZiYTYwMmU1LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDAxVDEzMDQ1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZmNzJkMDkyZDJmYjFiMjRhZmM4ODdlYTUwMDY1MzMyZTc3YWVlMTY0ODAyOTE1NjBjYmQ2OWU0N2ZjMzZlZDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Flckf3Jd-yXbU2jvjdiqwnPC7QNIlHJ4fD7N2bF5Z5I)

I think this might need my PR's to work correctly. https://github.com/cardstack/boxel/pull/1615 and https://github.com/cardstack/boxel/pull/1616. this reminds me of having an ID to a private card in your operator mode state url params, but you are logging in as a user that does not have access to such a card

**Comment by jurgenwerk** [comment]
Thanks @backspace for catching this, will check. Reviewers please off with the review until I rebase with newest changes from @habdelra's PRs once they are merged

**Comment by habdelra** [comment]
my hope is that at least as far as logging in goes, there will not have to be anything extra you need to do to support showing the login page after you merge my changes. moving the index HTML to the realm server should hopefully have simplified all that.

**Comment by backspace** [comment]
> Thanks @backspace for catching this, will check. Reviewers please off with the review until I rebase with newest changes from @habdelra's PRs once they are merged

I’m seeing the same behaviour [on this preview deployment](https://cs-7235-add-loading-state.boxel-host-preview.stack.cards/?operatorModeEnabled=true&operatorModeState=%7B%22stacks%22%3A%5B%5B%7B%22format%22%3A%22isolated%22%2C%22id%22%3A%22https%3A%2F%2Fcardstack.com%2Fbase%2Findex%22%7D%5D%5D%2C%22submode%22%3A%22interact%22%7D) so maybe it’s unrelated!

**Comment by habdelra** [comment]
> > Thanks @backspace for catching this, will check. Reviewers please off with the review until I rebase with newest changes from @habdelra's PRs once they are merged
> 
> I’m seeing the same behaviour [on this preview deployment](https://cs-7235-add-loading-state.boxel-host-preview.stack.cards/?operatorModeEnabled=true&operatorModeState=%7B%22stacks%22%3A%5B%5B%7B%22format%22%3A%22isolated%22%2C%22id%22%3A%22https%3A%2F%2Fcardstack.com%2Fbase%2Findex%22%7D%5D%5D%2C%22submode%22%3A%22interact%22%7D) so maybe it’s unrelated!

when you test this which credentials are you using? creds for staging? is this actually reproducible in one of our normal hosted envs? staging or production? or is it reproducible locally? I just wanna make sure we are not chasing ghosts.

**Comment by habdelra** [comment]
> > > Thanks @backspace for catching this, will check. Reviewers please off with the review until I rebase with newest changes from @habdelra's PRs once they are merged
> > 
> > 
> > I’m seeing the same behaviour [on this preview deployment](https://cs-7235-add-loading-state.boxel-host-preview.stack.cards/?operatorModeEnabled=true&operatorModeState=%7B%22stacks%22%3A%5B%5B%7B%22format%22%3A%22isolated%22%2C%22id%22%3A%22https%3A%2F%2Fcardstack.com%2Fbase%2Findex%22%7D%5D%5D%2C%22submode%22%3A%22interact%22%7D) so maybe it’s unrelated!
> 
> do you actually see this running locally? I wonder
> 
> > > Thanks @backspace for catching this, will check. Reviewers please off with the review until I rebase with newest changes from @habdelra's PRs once they are merged
> > 
> > 
> > I’m seeing the same behaviour [on this preview deployment](https://cs-7235-add-loading-state.boxel-host-preview.stack.cards/?operatorModeEnabled=true&operatorModeState=%7B%22stacks%22%3A%5B%5B%7B%22format%22%3A%22isolated%22%2C%22id%22%3A%22https%3A%2F%2Fcardstack.com%2Fbase%2Findex%22%7D%5D%5D%2C%22submode%22%3A%22interact%22%7D) so maybe it’s unrelated!
> 
> when you test this which credentials are you using? creds for staging? is this actually reproducible in one of our normal hosted envs? staging or production? or is it reproducible locally? I just wanna make sure we are not dealing with a red herring.

**Comment by backspace** [comment]
I don’t see it on staging, I haven’t tried running locally but will

ETA it doesn’t happen locally, sorry all!

**Comment by jurgenwerk** [comment]
In the [PR preview link](https://cs-7178-strip-out-guest-mode-as-a-front-end-feature.boxel-host-preview.stack.cards/) after login, something odd happens in the `start()` in `matrixService`, which then results in a logout. Here is the error thrown:

image

In the call stack I was able to trace this back to where we call `await this.matrixService.start({
                auth: auth1,
                refreshRoutes: true
            });` - perhaps this could be a  bug introduced in https://github.com/cardstack/boxel/pull/1615, @habdelra  ? 
            
I am not seeing this bug locally, also it doesn't look like it has been introduced in this PR as @buck already discovered

File: packages/host/app/routes/index.ts [hunk 1]
```diff
@@ -29,29 +32,103 @@ export default class Index extends Route<void> {
...
+      if (hostsOwnAssets) {
+        cardUrl = new URL(`/${cardPath}`, this.realm.defaultReadableRealm.path)
+          .href;
+      } else {
+        cardUrl = new URL(cardPath, window.location.origin).href;
+      }
+
+      let card = await this.fetchCard(cardUrl); // We are fetching the card here to make sure we have the canonical URL for the card
+
+      if (!card) {
```
Comment (jurgenwerk) [inline packages/host/app/routes/index.ts:75]: I have a struggle with a really weird bug here - this situation (sometimes!) happens when logging out and then attempting to view a card, and logging back in (which refreshes the route and reaches this point)... `fetchCard` will fail with a 401 even when logged in and having permission. I think there could be some kind of a race condition with `reauthenticate` that happens in the auth middleware which might be still using old data from the previous login (?)

File: packages/host/app/routes/index.ts [hunk 2]
```diff
@@ -29,29 +32,103 @@ export default class Index extends Route<void> {
...
+        // TODO: fix the root cause instead of this hack
+        await new Promise((resolve) => setTimeout(resolve, 500)); // Wait for a bit to so that the realm session is established
+        card = await this.fetchCard(cardUrl);
+      }
+
+      let canonicalCardUrl = card?.id;
+
+      if (!canonicalCardUrl) {
+        // TODO: what to do if card is not found, or not accessible?
+        alert(`Card not found: ${cardUrl}`);
```
Comment (jurgenwerk) [inline packages/host/app/routes/index.ts:89]: Currently this alerts and then opens the card chooser. Should we design a better notification?

**Comment by jurgenwerk** [comment]
I have deployed and tested this on staging and the new functionality works ok. I'm debugging through [that auth bug](https://github.com/cardstack/boxel/pull/1610/files#r1784426764) I noted in the comments, the rest should be ok to review now.

**Comment by habdelra** [comment]
> In the [PR preview link](https://cs-7178-strip-out-guest-mode-as-a-front-end-feature.boxel-host-preview.stack.cards/) after login, something odd happens in the `start()` in `matrixService`, which then results in a logout. Here is the error thrown:
> 
> image
> In the call stack I was able to trace this back to where we call `await this.matrixService.start({ auth: auth1, refreshRoutes: true });` - perhaps this could be a bug introduced in #1615, @habdelra ?
> 
> I am not seeing this bug locally, also it doesn't look like it has been introduced in this PR as @buck already discovered

buck actually reported this error before  #1615 was merged. it actually looks like you are getting HTML back from the synapse server instead of JSON. I think if you look closely you are getting some kind of AWS outage HTML. I think this is all down to the fact that the preview environment isn't really a fully supported one. Perhaps the preview environment is not configured with the correct matrix server URL. Probably we are just seeing this now because until very recently we had no deployed private realms, and all the realms up to now never required realm authentication to access.

File: packages/host/app/routes/index.ts (hunk 1 above)
Comment (habdelra) [reply packages/host/app/routes/index.ts:75]: So @lukemelia just addressed that particular scenario in his refactor of the authed fetch that we use (https://github.com/cardstack/boxel/pull/1601). Maybe merging that in will address this issue?

File: packages/host/app/routes/index.ts (hunk 2 above)
Comment (habdelra) [reply packages/host/app/routes/index.ts:89]: totally, we should have an official 404 page. Let's make a ticket to design this

File: packages/host/app/routes/index.ts [hunk 3]
```diff
@@ -29,29 +32,103 @@ export default class Index extends Route<void> {
...
+          },
+        ],
+      ];
+    }
+    if (!operatorModeState) {
       this.router.transitionTo('index', {
         queryParams: {
-          workspaceChooserOpened: 'true',
+          cardPath: undefined,
+          workspaceChooserOpened: stacks.length === 0,
```
Comment (habdelra) [inline packages/host/app/routes/index.ts:111]: love it!

File: packages/host/app/routes/index.ts [hunk 4]
```diff
@@ -29,29 +32,103 @@ export default class Index extends Route<void> {
...
+      if (this.operatorModeStateService.serialize() === operatorModeState) {
+        // If the operator mode state in the query param is the same as the one we have in memory,
+        // we don't want to restore it again, because it will lead to rerendering of the stack items, which can
+        // bring various annoyances, e.g reloading of the items in the index card.
+        // We will reach this point when the user manipulates the stack and the operator state service will set the
+        // query param, which will trigger a refresh of the model, which will call the model hook again.
+        // The model refresh happens automatically because we have operatorModeState: { refreshModel: true } in the queryParams.
+        // We have that because we want to support back-forward navigation in operator mode.
+        return;
+      }
```
Comment (habdelra) [inline packages/host/app/routes/index.ts:131]: does this fix CS-7247?

**Review by habdelra — APPROVED** [review]
looks fantastic! love all the deleted lines!

**Comment by habdelra** [comment]
also now that there is no more guest mode can we get rid of the basic auth on the realm server? (@lukemelia )

**Comment by lukemelia** [comment]
I've retargeted #1601 to this PR and plan to merge into this once it is green.

Update: merged.

**Comment by lukemelia** [comment]
> also now that there is no more guest mode can we get rid of the basic auth on the realm server? (@lukemelia )

Yes, I think we can and should do this in a separate PR

**Pull request merged**